segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/ConcurrencyManager.java

```diff
@@ -0,0 +1,209 @@
+                resetSnapshot(time);
+            } else if (elapsedMillis >= UPDATE_FREQUENCY_MILLIS) {
+                // Enough time elapsed since the last update; update the snapshot.
+                updateSnapshot(time, elapsedMillis);
+            }
+
+            return this.lastSnapshot.parallelism;
+        }
+    }
+
+    /**
+     * Updates the current snapshot based on the current data.
+     *
+     * @param time          The current time, in millis.
+     * @param elapsedMillis The elapsed time, in millis, since the last time this method ran.
+     */
+    @GuardedBy("snapshotLock")
+    private void updateSnapshot(long time, long elapsedMillis) {
+        final int recentCount;
+        final long recentLength;
+        synchronized (this) {
+            recentCount = this.recentWriteCount;
+            recentLength = this.recentTotalWrittenLength;
+            this.recentWriteCount = 0;
+            this.recentTotalWrittenLength = 0;
+        }
+
+        final double recentThroughput = (double) recentLength / elapsedMillis;
+        final double recentFillRatio = WriteQueue.calculateFillRatio(recentLength, recentCount);
+        final double throughputDifference = recentThroughput - this.lastSnapshot.throughput;
+
+        int parallelism = this.lastSnapshot.parallelism;
+        if (Math.abs(throughputDifference / this.lastSnapshot.throughput) >= SIGNIFICANT_DIFFERENCE) {
+            // Throughput changed significantly.
+            // But first, check for a special case. If both Throughput and Fill Ratio are decreasing, then most likely the
+            // former is a result of the latter. As such, don't do anything, in that case.
+            boolean fillRatioDecreased = (recentFillRatio - this.lastSnapshot.fillRatio) <= -SIGNIFICANT_DIFFERENCE;
+            if (throughputDifference > 0 || !fillRatioDecreased) {
+                // Either the throughput increased or the Fill Ratio did not decrease significantly.
+                // Change the parallelism in the same direction as Throughput.
+                // Update the degree of parallelism, but make sure we don't exceed the given bounds.
+                parallelism = MathHelpers.minMax(parallelism + (int) Math.signum(throughputDifference), this.minParallelism, this.maxParallelism);
+            }
+        } else if (parallelism == this.minParallelism) {
+            parallelism = Math.min(this.maxParallelism, parallelism + 1);
+        }
+
+        int age = this.lastSnapshot.age;
+        if (parallelism == this.lastSnapshot.parallelism && age >= MAX_STAGNATION_AGE) {
+            // If we have been stuck at this degree of parallelism for too long, nudge the parallelism up or down by a bit.
+            parallelism = MathHelpers.minMax(parallelism + (parallelism == this.maxParallelism ? -1 : 1), this.minParallelism, this.maxParallelism);
+        }
+
+        if (parallelism != this.lastSnapshot.parallelism) {
```

Comment:
If we make the suggested change above, then it looks like we don't need the snapshot age any longer.

Reply:
removed too

Reply:
Given our latest conversation on slack, there is still a good reason for this.